File: 1_problemSolvingPatterns/5_cyclic_sort/1_mycsharp/6a_FirstKMissingPositiveNumbers.cs
/*
[Hard but Starred]
Given an unsorted array containing numbers and a number ‘k’, 
find the first ‘k’ missing positive numbers in the array.

Example 1:

Input: [3, -1, 4, 5, 5], k=3
Output: [1, 2, 6]
Explanation: The smallest missing positive numbers are 1, 2 and 6.

Example 2:

Input: [2, 3, 4], k=3
Output: [1, 5, 6]
Explanation: The smallest missing positive numbers are 1, 5 and 6.

Example 3:

Input: [-2, -3, 4], k=2
Output: [1, 2]
Explanation: The smallest missing positive numbers are 1 and 2.


=== Solution
- Sort numbers in-place
- Find missing numbers using their indexes
- Also, separately keep the actual values that are in incorrect positions, so that when the K length is not met yet and want to find missing numbers beyond the array, it can be used for checking
- Find mising numbers using indexes beyond the array
*/


using System;
namespace _1_mycsharp
{
	public class _a_FirstKMissingPositiveNumbers
	{
		/// <summary>
		/// Find the first K positive missing numbers
		/// </summary>
		/// <param name="numbers">input arrays</param>
		/// <param name="k">number of missing numbers</param>
		public static void FindTheFirstKPositiveMissingNumbers(int[] numbers, int k)
		{
			int indexPointer = 0;
			while (indexPointer < numbers.Length)
			{
				// place numbers at their right locations
				var currentNumberAndIndex = numbers[indexPointer];
				if (currentNumberAndIndex > 0 && currentNumberAndIndex <= numbers.Length && currentNumberAndIndex != numbers[currentNumberAndIndex - 1])
				{
					// swap
					var temp = numbers[indexPointer];
					numbers[indexPointer] = numbers[currentNumberAndIndex - 1];
					numbers[currentNumberAndIndex - 1] = temp;
				}
				else
				{
					indexPointer++;
				}
			}
			Console.WriteLine($"Sorted numbers: {String.Join(", ", numbers)}");

			List<int> missingNumbers = new List<int>(); // using stored index
			List<int> extraNumbersOutOfRange = new List<int>();  // using actual corresponding values of missing numbers, that are out of 1-n range

			for (int i = 0; i < numbers.Length; i++)
			{
				// if missing number is not up to K, and
				// if a number is not placed in their right slot,
				// store their missing number (using index + 1) and current value (using value)
				if (missingNumbers.Count < k && numbers[i] != (i + 1))
				{
					missingNumbers.Add(i + 1);
					extraNumbersOutOfRange.Add(numbers[i]);	// store out of range numbers cus will be used to determine whether a number (that is beyond array bound) is missed or not
				}
			}
			Console.WriteLine($"First K +ve missin' numbers (init): {String.Join(", ", missingNumbers)}...Extra Numbers: {String.Join(", ", extraNumbersOutOfRange)}");


			// fill rest of missing numbers if not up to the expected K number
			// start outside the input array range boarder by increasing its length by 1
			int increamentalIndex = 1;
			while (missingNumbers.Count < k)
			{
				int missingNumberBeyondArrayBoarder = numbers.Length + increamentalIndex;

				// before marking as missedNumber, check whether it's found in the outOfRange list, if found, skip, it not a missing number since it's found in the original input array
				if (extraNumbersOutOfRange.Contains(missingNumberBeyondArrayBoarder))
				{
                    increamentalIndex += 1;
                    continue; // to next loop
				}
				missingNumbers.Add(missingNumberBeyondArrayBoarder);
                increamentalIndex += 1;
            }
            Console.WriteLine($"First K +ve missin' numbers (final): {String.Join(", ", missingNumbers)}");
		}


        /// <summary>
        /// Improve to make it more efficient by introducing Set data structure + positive-check when populating out-of-range extra numbers
		/// Clean comments
		/// Extra Numbers console differs and optimised
        /// </summary>
        /// <param name="numbers"></param>
        /// <param name="k"></param>
        public static void FindTheFirstKPositiveMissingNumbersImproved(int[] numbers, int k)
        {
            int indexPointer = 0;
            while (indexPointer < numbers.Length)
            {
                var currentNumberAndIndex = numbers[indexPointer];
                if (currentNumberAndIndex > 0 && currentNumberAndIndex <= numbers.Length && currentNumberAndIndex != numbers[currentNumberAndIndex - 1])
                {
                    // swap
                    var temp = numbers[indexPointer];
                    numbers[indexPointer] = numbers[currentNumberAndIndex - 1];
                    numbers[currentNumberAndIndex - 1] = temp;
                }
                else
                {
                    indexPointer++;
                }
            }
            Console.WriteLine($"Sorted numbers: {String.Join(", ", numbers)}");

            List<int> missingNumbers = new List<int>();
            HashSet<int> extraNumbersOutOfRange = new HashSet<int>();   // to prevent duplicate

            for (int i = 0; i < numbers.Length; i++)
            {
                if (missingNumbers.Count < k && numbers[i] != (i + 1))
                {
                    missingNumbers.Add(i + 1);
                    if (numbers[i] > 0)		// keep only positives
                    {
                        extraNumbersOutOfRange.Add(numbers[i]);
                    }
                }
            }
            Console.WriteLine($"First K +ve missin' numbers (init): {String.Join(", ", missingNumbers)}...Extra Numbers: {String.Join(", ", extraNumbersOutOfRange)}");

            int increamentalIndex = 1;
            while (missingNumbers.Count < k)
            {
                int missingNumberBeyondArrayBoarder = numbers.Length + increamentalIndex;

				if (!extraNumbersOutOfRange.Contains(missingNumberBeyondArrayBoarder))
                {
                    missingNumbers.Add(missingNumberBeyondArrayBoarder);
                }
                increamentalIndex += 1;
            }
            Console.WriteLine($"First K +ve missin' numbers (final): {String.Join(", ", missingNumbers)}");
        }

    }
}


/*

Time complexity #
The time complexity is O(n+k), as the last two for loops will run for O(n) and O(k) times respectively.

Space complexity #
The algorithm needs O(k) space to store the 'extraNumbers'

*/

File: 1_problemSolvingPatterns/5_cyclic_sort/1_mycsharp/6a_FirstKMissingPositiveNumbers.js
/*
[Hard but Starred]
Given an unsorted array containing numbers and a number ‘k’, 
find the first ‘k’ missing positive numbers in the array.

Example 1:

Input: [3, -1, 4, 5, 5], k=3
Output: [1, 2, 6]
Explanation: The smallest missing positive numbers are 1, 2 and 6.

Example 2:

Input: [2, 3, 4], k=3
Output: [1, 5, 6]
Explanation: The smallest missing positive numbers are 1 only. 
But since the result array size need to be same as k, add two other elements greater than max number in the original array, ie 5 and 6

Example 3:

Input: [-2, -3, 4], k=2
Output: [1, 2]
Explanation: The smallest missing positive numbers are 1 and 2.
*/



const x = smallestKMissingPositiveNumbers = (arr, numOfTimes) => {
    // cyclic sort
    // go through the array from start to finish
    let index = 0;

    while (index < arr.length) {
        const curentNumber = arr[index];
        const correctNumber = arr[curentNumber - 1];

        if(curentNumber > 0 && curentNumber < arr.length && curentNumber != correctNumber){    // considering only positive numbers in within range
            [arr[index], arr[curentNumber - 1]] = [arr[curentNumber - 1], arr[index]];
        }
        else {
            index++;
        }
    }
    
    let missingValues = [];

    for (let index = 0; index < arr.length; index++) {
        if(arr[index] != (index + 1)){
            missingValues.push(index + 1)     // missing numbers implies INDEX
        }
    }
    console.log("missingValues:", missingValues);

    // == adding extra array values to meet K size
    let arrLength = arr.length;    // values beyond array to add at L#59
    let increaseArrLengthValueBy1 = 1;

    while(missingValues.length < numOfTimes)    // fill the result bucket with big numbers until it's up to the k size required
    {
        let nextValueToAdd = arrLength + increaseArrLengthValueBy1;
        if(!arr.includes(nextValueToAdd)){   // add number that is not already in the original array
            missingValues.push(nextValueToAdd)
        }
        increaseArrLengthValueBy1 ++;
    }

    return missingValues;

    // my algo======
    //      if an element is not at expected place based on its index
    //          swap element to move to correct place
    // once all is done, array is sorted

    // storage 
    // now, go through the sorted array
    //      if element is not expected place, store its index in storage

    // storage length is not up to k yet,
    //      add next number(beyond array size) to storage, but make sure it's not found in array

    // return storage
}

console.log("final missingValues::", x([3, -1, 4, 5, 5], 3));


/*

Tip so far: 

YOU THINK... => DO...
1. unsorted input + occurrence                                      => frequency counter map
2. sorted input                                                     => binary search, two-pointer 
3. sub continuous array                                             => sliding window  
4. cyclic + linkedlist midpoint                                     => fast n slow (tortoise and hare) 
5. inputs range from 1/0 to n/(n-1) or small overrange difference   => cyclic

*/